### PR TITLE
Relative adsorption, interfacial enrichment, and interface thickness.

### PR DESCRIPTION
**Pull request by RolfStierle**
…(90-10; just in Python interface, otherwise general) implemented.

### CODE REVIEW

File: feos-dft/src/interface/mod.rs [hunk 1]
```diff
@@ -196,6 +196,142 @@ impl<U: EosUnit, F: HelmholtzEnergyFunctional> PlanarInterface<U, F> {
...
+
+        // Calculate interfacial enrichment E_i
+        let enrichment = Array1::from_shape_fn(s[0], |i| {
+            (*(self
+                .profile
+                .density
+                .index_axis(Axis_nd(0), i)
+                .to_owned()
+                .to_reduced(U::reference_density())
+                .unwrap()
```
Comment (g-bauer) [inline feos-dft/src/interface/mod.rs:252]: Replace this `unwrap` with `?`

File: feos-dft/src/interface/mod.rs [hunk 2]
```diff
@@ -196,6 +196,142 @@ impl<U: EosUnit, F: HelmholtzEnergyFunctional> PlanarInterface<U, F> {
...
+        } else {
+            (limits.1, limits.0)
+        };
+
+        if limit_upper >= 1.0 || limit_upper.is_sign_negative() {
+            panic!("Upper limit 'l' of interface thickness needs to satisfy 0 < l < 1.");
+        }
+        if limit_lower >= 1.0 || limit_lower.is_sign_negative() {
+            panic!("Lower limit 'l' of interface thickness needs to satisfy 0 < l < 1.");
+        }
```
Comment (g-bauer) [inline feos-dft/src/interface/mod.rs:286]: Maybe use `Result` instead of panic?

File: feos-dft/src/python/interface/mod.rs [hunk 3]
```diff
@@ -114,5 +114,48 @@ macro_rules! impl_planar_interface {
...
+            /// Calculates the Gibbs' relative adsorption of component `i' with
+            /// respect to `j': \Gamma_i^(j)
+            ///
+            /// Returns
+            /// -------
+            /// SIArray2
+            ///
+            #[pyo3(text_signature = "($self)")]
+            fn relative_adsorption(&self) -> PyResult<PySIArray2> {
+                Ok(self.0.relative_adsorption()?.into())
```
Comment (g-bauer) [inline feos-dft/src/python/interface/mod.rs:129]: Is this `into` needed?

File: feos-dft/src/python/interface/mod.rs [hunk 4]
```diff
@@ -114,5 +114,48 @@ macro_rules! impl_planar_interface {
...
+            ///     The bulk phase equilibrium.
+            ///
+            /// Returns
+            /// -------
+            /// SINumber
+            ///
+            #[pyo3(text_signature = "($self)")]
+            fn interfacial_thickness(&self) -> PyResult<PySINumber> {
+                let limits = (0.9, 0.1);
+                Ok(self.0.interfacial_thickness(limits)?.into())
```
Comment (g-bauer) [inline feos-dft/src/python/interface/mod.rs:157]: idem

File: tests/pcsaft/dft.rs [hunk 5]
```diff
@@ -354,3 +354,40 @@ fn test_entropy_bulk_values() -> Result<(), Box<dyn Error>> {
     );
     Ok(())
 }
+
```
Comment (g-bauer) [inline tests/pcsaft/dft.rs:357]: Remove test or un-comment?

**Review by g-bauer — CHANGES_REQUESTED** [review]

File: feos-dft/src/interface/mod.rs [hunk 6]
```diff
@@ -196,6 +196,142 @@ impl<U: EosUnit, F: HelmholtzEnergyFunctional> PlanarInterface<U, F> {
...
+            if i == j {
+                0.0
+            } else {
+                (-(rho_l.get(i) - rho_v.get(i))
+                    * ((&self.profile.density.index_axis(Axis_nd(0), j) - rho_l.get(j))
+                        / (rho_l.get(j) - rho_v.get(j))
+                        - (&self.profile.density.index_axis(Axis_nd(0), i) - rho_l.get(i))
+                            / (rho_l.get(i) - rho_v.get(i)))
+                    .integrate(&self.profile.grid.integration_weights_unit()))
+                .to_reduced(U::reference_density() * U::reference_length())
```
Comment (prehner) [inline feos-dft/src/interface/mod.rs:224]: There is no need to used reduced units here. You can use `QuantityArray2::from_shape_fn`

File: feos-dft/src/interface/mod.rs [hunk 7]
```diff
@@ -196,6 +196,142 @@ impl<U: EosUnit, F: HelmholtzEnergyFunctional> PlanarInterface<U, F> {
...
+        if limit_lower >= 1.0 || limit_lower.is_sign_negative() {
+            panic!("Lower limit 'l' of interface thickness needs to satisfy 0 < l < 1.");
+        }
+
+        // Get the densities in the liquid and in the vapor phase
+        let rho_l = if rho.get(0) > rho.get(s[1] - 1) {
+            rho[0]
+        } else {
+            rho[s[1] - 1]
+        };
```
Comment (prehner) [inline feos-dft/src/interface/mod.rs:293]: Can this be replaced with `max` and `min?`

File: feos-dft/src/interface/mod.rs [hunk 8]
```diff
@@ -230,6 +366,8 @@ impl<U: EosUnit, F: HelmholtzEnergyFunctional> PlanarInterface<U, F> {
     }
 }
 
+impl<U: EosUnit, F: HelmholtzEnergyFunctional> PlanarInterface<U, F> {}
```
Comment (prehner) [inline feos-dft/src/interface/mod.rs:369]: ??

File: feos-dft/src/python/interface/mod.rs [hunk 9]
```diff
@@ -114,5 +114,48 @@ macro_rules! impl_planar_interface {
...
+            /// Parameters
+            /// ----------
+            /// limits : tuple('float', 'float')
+            ///     The bulk phase equilibrium.
+            ///
+            /// Returns
+            /// -------
+            /// SINumber
+            ///
+            #[pyo3(text_signature = "($self)")]
```
Comment (prehner) [inline feos-dft/src/python/interface/mod.rs:154]: Fix docstring or (preferred) use the same interface for Python and Rust.

File: feos-dft/src/python/interface/mod.rs (hunk 3 above)
Comment (RolfStierle) [reply feos-dft/src/python/interface/mod.rs:129]: Does not compile otherwise.

File: feos-dft/src/python/interface/mod.rs (hunk 4 above)
Comment (RolfStierle) [reply feos-dft/src/python/interface/mod.rs:157]: Does not compile otherwise.

File: tests/pcsaft/dft.rs (hunk 5 above)
Comment (RolfStierle) [reply tests/pcsaft/dft.rs:357]: Done.

File: feos-dft/src/interface/mod.rs (hunk 7 above)
Comment (RolfStierle) [reply feos-dft/src/interface/mod.rs:293]: Yes, thank you.

File: feos-dft/src/interface/mod.rs (hunk 6 above)
Comment (RolfStierle) [reply feos-dft/src/interface/mod.rs:224]: Thanks, that simplifies it a bit.

File: feos-dft/src/interface/mod.rs (hunk 1 above)
Comment (RolfStierle) [reply feos-dft/src/interface/mod.rs]: Replaced unwraps as far as I was able to.

**Comment by prehner** [comment]
Thank you for the revisions! It looks like a nice addition. Can you remove the commented out lines of code? I think without context they are not helpful. Finally please add a changelog entry to the changelog of feos-dft (careful, feos and feos-dft have their own changelogs).

**Review by prehner — APPROVED** [review]

**Review by g-bauer — APPROVED** [review]

**Pull request merged**